> with
> 
> ```
> fmw@fmw-xps-13:~/git/robot_ws/src/cob_command_tools (refactor_hz_monitor)$ rosparam set /hz_monitor/topics [/test]
> fmw@fmw-xps-13:~/git/robot_ws/src/cob_command_tools (refactor_hz_monitor)$ rostopic pub /test std_msgs/String "data: 'asd'" -r10
> fmw@fmw-xps-13:~/git/robot_ws/src/cob_command_tools (refactor_hz_monitor)$ rosrun cob_monitoring hz_monitor.py 
> ```
> 
> and then stopping the publisher, I get the following error:
> 
> ```
> [INFO] [1581088523.227710]: subscribed to [/test]
> Exception in thread Thread-4:
> Traceback (most recent call last):
>   File "/usr/lib/python2.7/threading.py", line 801, in __bootstrap_inner
>     self.run()
>   File "/opt/ros/kinetic/lib/python2.7/dist-packages/rospy/timer.py", line 235, in run
>     self._callback(TimerEvent(last_expected, last_real, current_expected, current_real, last_duration))
>   File "/home/fmw/git/robot_ws/src/cob_command_tools/cob_monitoring/src/hz_monitor.py", line 131, in publish_diagnostics
>     and not (event.current_real.to_sec() < rt.msg_tn + 1/self.hz):  # condition to prevent WARN for topics with hz<diagnostics_rate
> TypeError: unsupported operand type(s) for /: 'int' and 'NoneType'
> ```

you did not specify a desired `hz` rate and now `hzerror` epsilon for the topic...this defaulted to `None`...which of course cannot be used in a division...
I now made this a mandatory parameter and now the node exits with a proper message

--- a/cob_monitoring/src/hz_monitor.py
+++ b/cob_monitoring/src/hz_monitor.py
@@ -32,9 +32,9 @@ class HzTest():
             # topic to test
             self.topics = rospy.get_param('~topics')
             # expected publishing rate
-            self.hz = rospy.get_param('~hz', None)
+            self.hz = rospy.get_param('~hz')
             # margin of error allowed
-            self.hzerror = rospy.get_param('~hzerror', None)
+            self.hzerror = rospy.get_param('~hzerror')
             # length of test
             self.window_size = float(rospy.get_param('~window_size', 100))
             # name for diagnostic message

--- a/cob_monitoring/src/hz_monitor.py
+++ b/cob_monitoring/src/hz_monitor.py
@@ -76,16 +76,20 @@ class HzTest():
 
     def publish_diagnostics(self, event):
         # set desired rates
-        if self.hzerror:
-            if isinstance(self.hzerror, float) or isinstance(self.hzerror, int):
-                min_rate = self.hz - self.hzerror
-                max_rate = self.hz + self.hzerror
-            else:
-                rospy.logerr("hzerror not float or int")
+        if isinstance(self.hzerror, float) or isinstance(self.hzerror, int):
+            if self.hzerror < 0:
+                rospy.logerr("hzerror cannot be negative")
                 sys.exit(1)
+            min_rate = self.hz - self.hzerror
+            max_rate = self.hz + self.hzerror
+            if min_rate < 0 or max_rate < 0:
+                rospy.logerr("min_rate/max_rate cannot be negative")
+                sys.exit(1)
+            min_duration = 1/min_rate if min_rate > 0 else float('inf')
+            max_duration = 1/max_rate if max_rate > 0 else float('inf')
         else:
-            min_rate = None
-            max_rate = None
+            rospy.logerr("hzerror not float or int")
+            sys.exit(1)
 
         # create diagnostic message
         array = DiagnosticArray()
@@ -128,7 +132,7 @@ class HzTest():
                 rates.append(0.0)
                 consolidated_error_messages["never received message for topics"].append(topic)
             elif rt.msg_tn == rt.last_printed_tn \
-                 and not (event.current_real.to_sec() < rt.msg_tn + 1/self.hz):  # condition to prevent WARN for topics with hz<diagnostics_rate
+                 and not (event.current_real.to_sec() < rt.msg_tn + min_duration):  # condition to prevent WARN for topics with hz<diagnostics_rate, allow 1/min_rte to pass before reporting error
                 hz_status.level = DiagnosticStatus.ERROR
                 rates.append(0.0)
                 consolidated_error_messages["no messages anymore for topics"].append(topic)

--- a/cob_monitoring/src/hz_monitor.py
+++ b/cob_monitoring/src/hz_monitor.py
@@ -40,7 +40,7 @@ class HzTest():
             # name for diagnostic message
             self.diagnostics_name = rospy.get_param('~diagnostics_name',"")
             self.diagnostics_name = self.diagnostics_name.replace('/','_')
-            self.diagnostics_rate = 1.0
+            self.diagnostics_period = rospy.Duration(1.0)
         except KeyError as e:
             rospy.logerr('hztest not initialized properly. Parameter [{}] not set. debug[{}] debug[{}]'.format(e, rospy.get_caller_id(), rospy.resolve_name(e.args[0])))
             sys.exit(1)
@@ -48,7 +48,7 @@ class HzTest():
         self.rt_HZ_store = dict()
         self.missing_topics = copy.deepcopy(self.topics)
         self.pub_diagnostics = rospy.Publisher('~diagnostics', DiagnosticArray, queue_size = 1)
-        self.timer_diagnostics = rospy.Timer(rospy.Duration(self.diagnostics_rate), self.publish_diagnostics)
+        self.timer_diagnostics = rospy.Timer(self.diagnostics_period, self.publish_diagnostics)
 
 
     def run(self):
@@ -62,7 +62,7 @@ class HzTest():
 
             try:
                 rospy.logdebug("hz monitor is waiting for type of topics {}.".format(self.missing_topics))
-                rospy.Rate(self.diagnostics_rate).sleep()
+                rospy.sleep(1.0)
             except rospy.exceptions.ROSInterruptException:
                 pass
 

